cql3/restrictions/statement_restrictions.cc

```diff
@@ -1003,5 +1006,725 @@ ::shared_ptr<single_column_restriction> single_column_restriction::LIKE::apply_t
+    } else if (auto mkr = dynamic_cast<lists::marker*>(&rhs)) {
+        auto multi = dynamic_pointer_cast<multi_item_terminal>(mkr->bind(bag.options));
+        if (multi) {
+            if (cvs.size() != 1) {
+                throw std::logic_error("too many columns for lists::marker in is_one_of");
+            }
+            return boost::algorithm::any_of(multi->get_elements(), [&] (const bytes_opt& b) {
+                    return equal(b, cvs[0], bag);
+                });
+        }
+    } else if (auto mkr = dynamic_cast<tuples::in_marker*>(&rhs)) {
+        const auto val = dynamic_pointer_cast<tuples::in_value>(mkr->bind(bag.options));
+        if (!val) {
+            return false;
+        }
+        return boost::algorithm::any_of(val->get_split_values(), [&] (const std::vector<bytes_opt>& el) {
+                return boost::equal(cvs, el, [&] (const column_value& c, const bytes_opt& b) {
+                    return equal(b, c, bag);
+                });
+            });
+    }
+    throw std::logic_error("unexpected term type in is_one_of");
+}
+
+/// True iff op means bnd type of bound.
+bool matches(const operator_type* op, statements::bound bnd) {
+    static const std::vector<std::vector<const operator_type*>> operators{
+        {&operator_type::EQ, &operator_type::GT, &operator_type::GTE}, // These mean a lower bound.
+        {&operator_type::EQ, &operator_type::LT, &operator_type::LTE}, // These mean an upper bound.
+    };
+    const auto zero_if_lower_one_if_upper = get_idx(bnd);
+    return boost::algorithm::any_of_equal(operators[zero_if_lower_one_if_upper], op);
+}
+
+const value_set empty_value_set = value_list{};
+const value_set unbounded_value_set = value_interval{};
+
+struct intersection_visitor {
+    value_set operator()(const value_list& a, const value_list& b) const {
+        value_list common;
+        boost::set_intersection(a, b, back_inserter(common));
+        return std::move(common);
+    }
+
+    value_set operator()(const value_interval& a, const value_list& b) const {
+        const auto common = b | filtered([&] (const bytes& el) { return a.includes(el); });
+        return value_list(common.begin(), common.end());
+    }
+
+    value_set operator()(const value_list& a, const value_interval& b) const {
+        return (*this)(b, a);
+    }
+
+    value_set operator()(const value_interval& a, const value_interval& b) const {
+        const auto newlb = std::max(a.lb, b.lb); // nullopt is less than a value, so this works.
+        const auto newub = std::min(
+                a.ub, b.ub, [] (const std::optional<upper_bound>& x, const std::optional<upper_bound>& y) {
+                    if (x == y || !x) {
+                        return false; // Consider nullopt greater than a value, which becomes newub.
+                    } else  {
+                        return !y || y->includes(x->value);
+                    }
+                });
+        return value_interval{newlb, newub};
```

Comment:
With `typename Comparator`. ;)  Are you envisioning passing nonwrapping_range and its comparator everywhere I pass value_interval today?  Why are two parameters better than one?

Reply:
To avoid duplicating nonwrapping_range, and to avoid copying static data into temporaries, creating more work. We are very far from the point where we have complete separation between static and dynamic data, but let's try not to make it worse.

But I'm not thrilled about using this value_set everywhere. It looks like a lot of runtime work (slow), and also like it is fragile wrt grammar expansion, since simple things like a=b cannot be expressed using a set of values and intervals.

Reply:
As mentioned elsewhere, this will never encounter `a=b`.  I'll try to reuse nonwrapping_range as the return type and grab the comparator from each conjunction element as I go along.

Reply:
OK, I think I understand the upper/lower bound confusion.  A token value represents the entire partition, with potentially multiple rows.  Naturally, such partition will have a beginning and an end.  For `token() < 123`, the correct bound is the end of the partition whose token is 123, exclusive; but for `token() <= 123`, the correct bound is the beginning of that partition, inclusive.

Reply:
> So I've been struggling to switch to nonwrapping_range for some time, and I think I can make it work. There are, however, a few features of nonwrapping_range I find incomprehensible:
> 
> * Its methods choose to represent an empty range as nullopt, rather than a range with equal, exclusive bounds; for instance, the intersection() method returns a different type than the one it takes in, making composition very unpleasant.

It can't just invent a T value for the bound.

However, I agree it would be nicer to be able to support an empty range internally, with an extra boolean flag.


> * In several places, it suggests that an exclusive upper bound is really a lower bound, and vice versa (eg, [here](https://sourcegraph.com/github.com/scylladb/scylla@5105e9f5e1f1a56f5a41ca07e9b9c4329203be3b/-/blob/range.hh#L650), [here]

No, the names of the functions called (std::lower_bound/upper_bound) confuse you. For an inclusive bound, we want to find the first element that is not less than the inclusive bound, and for an upper bound we want to find the first element that is greater than the exclusive bound, and that is what the code does.

(https://sourcegraph.com/github.com/scylladb/scylla@5105e9f5e1f1a56f5a41ca07e9b9c4329203be3b/-/blob/range.hh#L659), [here](https://sourcegraph.com/github.com/scylladb/scylla@5105e9f5e1f1a56f5a41ca07e9b9c4329203be3b/-/blob/cql3/restrictions/token_restriction.hh#L128)).
> * Token ranges [treat](https://sourcegraph.com/github.com/scylladb/scylla@5105e9f5e1f1a56f5a41ca07e9b9c4329203be3b/-/blob/cql3/restrictions/token_restriction.hh#L99) `token(...) < MINTOKEN` as if the user wrote `token(...) < MAXTOKEN`.  Besides baffling the reader, this complicates code reuse.

This doesn't look like a problem with ranges (which should be renamed to intervals to avoid conflict with the C++ concept of a range), but rather with the token restriction code.

> 
> Either I'm short of fully understanding nonwrapping_range or nonwrapping_range should improve independently of this work.

I agree with your first point, there should be a way to represent an empty range without a wrapper optional<>.

Reply:
Switched to using nonwrapping_range.